mlir/lib/Dialect/Arith/Transforms/IntRangeOptimizations.cpp

```diff
@@ -184,6 +189,232 @@ struct DeleteTrivialRem : public OpRewritePattern<RemOp> {
   DataFlowSolver &solver;
 };
 
+static Type checkArithType(Type type, unsigned targetBitwidth) {
+  type = getElementTypeOrSelf(type);
+  if (isa<IndexType>(type))
+    return type;
+
+  if (auto intType = dyn_cast<IntegerType>(type))
+    if (intType.getWidth() > targetBitwidth)
+      return type;
+
+  return nullptr;
+}
+
+static Type checkElementwiseOpType(Operation *op, unsigned targetBitwidth) {
+  if (op->getNumOperands() == 0 || op->getNumResults() == 0)
+    return nullptr;
+
+  Type type;
+  for (auto range :
+       {ValueRange(op->getOperands()), ValueRange(op->getResults())}) {
```

Comment:
`concat_range`? Or if that doesn't work, materialize the vector of values instead of adding an extra loop.

Reply:
I didn't know about `llvm::concat` actually, cool. But unfortunately is didn't work for `ValueRange` for some reason. And using a vector is suboptimal as it can cause additional allocations while loop cannot.

Reply:
concat was updated recently (~2 weeks ago) to handle proxy iterators -- could you check if it'd work now?

Reply:
yeah, it works now